persistence/src/commonMain/kotlin/com/quran/shared/persistence/repository/BookmarksRepositoryImpl.kt

```diff
@@ -0,0 +1,197 @@
+package com.quran.shared.persistence.repository
+
+import app.cash.sqldelight.coroutines.asFlow
+import com.quran.shared.persistence.Bookmarks
+import com.quran.shared.persistence.Bookmarks_mutations
+import com.quran.shared.persistence.QuranDatabase
+import com.quran.shared.persistence.model.Bookmark
+import com.quran.shared.persistence.model.BookmarkLocalMutation
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.map
+
+class BookmarksRepositoryImpl(
+    private val database: QuranDatabase
+) : BookmarkRepository {
+    override fun getAllBookmarks(): Flow<List<Bookmark>> {
+        val persistedFlow = database.bookmarksQueries.getBookmarks()
+            .asFlow()
+            .map { query ->
+                query.executeAsList().map { it.toBookmark() }
+            }
+        val mutatedFlow = database.bookmarks_mutationsQueries.getBookmarksMutations()
+            .asFlow()
+            .map { query ->
+                query.executeAsList().map { it.toBookmark() }
+            }
+
+        return persistedFlow.combine(mutatedFlow) { persistedBookmarks, mutatedBookmarks ->
+            val deletedRemoteIDs = mutatedBookmarks
+                .filter { it.localMutation == BookmarkLocalMutation.DELETED }
+                .mapNotNull { it.remoteId }
+                .toSet()
+
+            persistedBookmarks.filter { it.remoteId !in deletedRemoteIDs } +
+                    mutatedBookmarks.filter { it.remoteId !in deletedRemoteIDs }
+        }
+    }
+
+    override suspend fun addPageBookmark(page: Int) {
+        addBookmark(page, null, null)
+    }
+
+    override suspend fun addAyahBookmark(sura: Int, ayah: Int) {
+        addBookmark(null, sura, ayah)
+    }
+
+    private fun addBookmark(page: Int?, sura: Int?, ayah: Int?) {
+        val mutatedBookmarks = database.bookmarks_mutationsQueries
```

Comment:
if we do merge these, this becomes easy to just do in sql rather than 2 calls - i.e. a `INSERT OR REPLACE` query